Backport of:

From dfaec84fda2f126e956d840612f4bed1e7c9f457 Mon Sep 17 00:00:00 2001
From: Cristy <urban-warrior@imagemagick.org>
Date: Sun, 14 Feb 2016 10:40:23 -0500
Subject: [PATCH] Fix out of bound read in viff file handling

bug: https://github.com/ImageMagick/ImageMagick/issues/129
bug-ubuntu: https://bugs.launchpad.net/ubuntu/+source/imagemagick/+bug/1545183
bug-debian: https://bugs.debian.org/845212
origin: upstream, https://github.com/ImageMagick/ImageMagick/commit/76ac0460463c7f4eab8e58a5dd5cbb2bb012ccd3

(cherry picked from commit 76ac0460463c7f4eab8e58a5dd5cbb2bb012ccd3)
---
 coders/viff.c | 28 ++++++++++++++++++++++++++--
 1 file changed, 26 insertions(+), 2 deletions(-)

Index: imagemagick-6.6.9.7/coders/viff.c
===================================================================
--- imagemagick-6.6.9.7.orig/coders/viff.c	2016-11-29 13:47:46.593937419 -0500
+++ imagemagick-6.6.9.7/coders/viff.c	2016-11-29 13:47:46.589937372 -0500
@@ -137,6 +137,22 @@
 %    o exception: return any errors or warnings in this structure.
 %
 */
+
+static MagickBooleanType CheckMemoryOverflow(const size_t count,
+  const size_t quantum)
+{
+  size_t
+    size;
+
+  size=count*quantum;
+  if ((count == 0) || (quantum != (size/count)))
+    {
+      errno=ENOMEM;
+      return(MagickTrue);
+    }
+  return(MagickFalse);
+}
+
 static Image *ReadVIFFImage(const ImageInfo *image_info,
   ExceptionInfo *exception)
 {
@@ -489,9 +505,17 @@
       default: bytes_per_pixel=1; break;
     }
     if (viff_info.data_storage_type == VFF_TYP_BIT)
-      max_packets=((image->columns+7UL) >> 3UL)*image->rows;
+      {
+        if (CheckMemoryOverflow((image->columns+7UL) >> 3UL,image->rows) != MagickFalse)
+          ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+        max_packets=((image->columns+7UL) >> 3UL)*image->rows;
+      }
     else
-      max_packets=(size_t) (number_pixels*viff_info.number_data_bands);
+      {
+        if (CheckMemoryOverflow(number_pixels,viff_info.number_data_bands) != MagickFalse)
+          ThrowReaderException(ResourceLimitError,"MemoryAllocationFailed");
+        max_packets=(size_t) (number_pixels*viff_info.number_data_bands);
+      }
     viff_pixels=(unsigned char *) AcquireQuantumMemory(MagickMax(number_pixels,
       max_packets),bytes_per_pixel*sizeof(*viff_pixels));
     if (viff_pixels == (unsigned char *) NULL)
